server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java

```diff
@@ -0,0 +1,424 @@
+                dataStream.getName(),
+                DATA_STREAMS_AUTO_SHARDING_EXCLUDES_SETTING.getKey()
+            );
+            return NOT_APPLICABLE_RESULT;
+        }
+
+        if (writeIndexLoad == null) {
+            logger.debug(
+                "Data stream auto sharding service cannot compute the optimal number of shards for data stream [{}] as the write index "
+                    + "load is not available",
+                dataStream.getName()
+            );
+            return NOT_APPLICABLE_RESULT;
+        }
+        return innerCalculate(metadata, dataStream, writeIndexLoad, nowSupplier);
+    }
+
+    private AutoShardingResult innerCalculate(Metadata metadata, DataStream dataStream, double writeIndexLoad, LongSupplier nowSupplier) {
+        // increasing the number of shards is calculated solely based on the index load of the write index
+        IndexMetadata writeIndex = metadata.index(dataStream.getWriteIndex());
+        assert writeIndex != null : "the data stream write index must exist in the provided cluster metadata";
+        AutoShardingResult increaseShardsResult = getIncreaseShardsResult(dataStream, writeIndexLoad, nowSupplier, writeIndex);
+        if (increaseShardsResult != null) {
+            return increaseShardsResult;
+        }
+
+        // the way we handle the cool down when reducing the number of shards is concerned is a bit different as it involves checking the
+        // write load of more than just the write index.
+        // because we have to look at the maximum write load encountered within the configured cool down period this can involve
+        // checking the write load for older backing indices. if the cool down period hasn't lapsed we short circuit here and return
+        // NO_CHANGE_REQUIRED because we don't want our historic write load lookup to include backing indices that are older than the
+        // index that triggered the previous auto scaling event
+        // this would happen if say, we increased the number of shards one day ago, and there's 2 days left of the reduce number of shards
+        // cool down - looking back the entire cool down period (3 days) would include the previous couple backing indices, including the
+        // one before the index that triggered the increase number of shards event)
+        TimeValue remainingReduceShardsCooldown = getRemainingDecreaseShardsCooldown(metadata, dataStream);
+        if (remainingReduceShardsCooldown.equals(TimeValue.ZERO) == false) {
+            Index oldestBackingIndex = dataStream.getIndices().get(0);
+            IndexMetadata oldestIndexMeta = metadata.getIndexSafe(oldestBackingIndex);
+            logger.debug(
+                "data stream autosharding service recommends no change in the number of shards for data stream [{}]  "
+                    + "with previous autosharding event [{}] and oldest generation index [{}] being created on [{}] ",
```

Comment:
Should this log message include something about the reason being that the remaining reduce shards cooldown being `N` (which is > 0)?

Reply:
This is not a thing anymore with the restructuring of the statuses.